### PR TITLE
Fix issue 159

### PR DESCRIPTION
**Pull request by roigcarlo**
This PR implements #159 

- [x] Update documentation on new API
- [x] Update ABC description of the ABCEngine
- [ ] Add a template for testing Engines
- [x] Update api of the H5CUDS file

### CODE REVIEW

File: simphony/cuds/abc_modeling_engine.py [hunk 1]
```diff
@@ -22,185 +22,82 @@ class ABCModelingEngine(object):
...
-    def iter_lattices(self, names=None):
-        """ Returns an iterator over a subset or all of the lattices.
-
-        Parameters
-        ----------
-        names : sequence of str, optional
-            names of specific lattices to be iterated over. If names is not
-            given, then all lattices will be iterated over.
+        container :
+            A proxy of the dataset named ``name`` that is stored
```
Comment (itziakos) [inline simphony/cuds/abc_modeling_engine.py]: suggestion: I would avoid using the word `proxy` in the case. I have seen a number of cases were developers can get into a heated discussion on terms like proxy and adapter and which pattern we follow. I would suggest something like

```
container:
    An object implementing the CUDS container API providing access to the information stored in the engine.
```

Related links: (http://programmers.stackexchange.com/questions/201397/difference-between-the-adapter-pattern-and-the-proxy-pattern).

File: simphony/io/h5_cuds.py [hunk 2]
```diff
@@ -66,7 +70,125 @@ def close(self):
...
+        """
+        if name in self._get_child_names(self._root.particle):
+            return self._get_particles(name)
+        elif name in self._get_child_names(self._root.mesh):
+            return self._get_mesh(name)
+        elif name in self._get_child_names(self._root.lattice):
+            return self._get_lattice(name)
+        else:
+            raise ValueError(
+                'Container \'{n}\` does not exist'.format(n=name))
```
Comment (itziakos) [inline simphony/io/h5_cuds.py]: note: one can use `"`, so that an escape character is not necessary.
```python
"Container '{n}` does not exist".
```

**Comment by itziakos** [comment]
@roigcarlo, just some minor comments. Will you have time to add a template test?

**Comment by roigcarlo** [comment]
I will try, to have it before tomorrow

**Comment by roigcarlo** [comment]
@itziakos I will probably won't have this untill tomorrow at about 11:00 along with #177. I will reping you otherwise

File: simphony/io/h5_cuds.py [hunk 3]
```diff
@@ -66,7 +70,125 @@ def close(self):
...
         self._handle.close()
 
-    def add_particles(self, particles):
+    def add_dataset(self, container):
+        """Add a CUDS container
+
+        Parameters
+        ----------
+        container : {ABCMesh, ABCParticles, ABCLattice}
+            The CUDS container to add to the engine.
```
Comment (nathanfranklin) [inline simphony/io/h5_cuds.py]: "The CUDS container to add to the engine." ->  "The CUDS container to be added to the file." or "The CUDS container to be added."

File: simphony/io/h5_cuds.py [hunk 4]
```diff
@@ -66,7 +70,125 @@ def close(self):
...
+        Parameters
+        ----------
+        name: str
+            name of CUDS container to be retrieved.
+
+        Returns
+        -------
+        container :
+            A proxy of the dataset named ``name`` that is stored
+            internally in the Engine.
```
Comment (nathanfranklin) [inline simphony/io/h5_cuds.py]: Engine -> file

File: simphony/cuds/abc_modeling_engine.py [hunk 5]
```diff
@@ -22,185 +22,82 @@ class ABCModelingEngine(object):
...
-            get_particles for more information.
-
-        """
-        pass
-
-    @abstractmethod
-    def delete_lattice(self, name):
-        """Delete a lattice
+    def remove_dataset(self, name):
+        """ Remove a dataset from the internal
```
Comment (nathanfranklin) [inline simphony/cuds/abc_modeling_engine.py]: "from the internal" -> "from the engine"

**Comment by nathanfranklin** [comment]
@roigcarlo ,  looks good.  just left some docstring comments.  Additionaly,  `simphony/bench/cuds_file_bench.py` will need to be updated.

**Comment by roigcarlo** [comment]
@itziakos Sorry I accidentaly merged this. there is a way to undo?

**Comment by SGGgarcia** [comment]
@roigcarlo normally there is a "Revert" button in the bottom of the PR, just next to the merge operation, but I dont see it in this one, I don't know if its just me or if its hidden for any other reason

**Comment by itziakos** [comment]
The ``revert`` operation is active only for a small time.

**Comment by roigcarlo** [comment]
Problem is that I didn't press the merge button, it was a typo in the terminal command...

File: simphony/testing/abc_check_engine.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+from simphony.cuds.particles import Particle, Particles
+from simphony.cuds.mesh import Point, Mesh
+from simphony.cuds.lattice import Lattice
+from simphony.io.h5_cuds import H5CUDS
+from simphony.io.h5_mesh import H5Mesh
+from simphony.io.h5_particles import H5Particles
+from simphony.io.h5_lattice import H5Lattice
+
+
+class DatasetCudsCheck(object):
```
Comment (itziakos) [inline simphony/testing/abc_check_engine.py]: Please rename to `CheckEngine`

File: simphony/testing/abc_check_engine.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+        """
+
+    @abc.abstractmethod
+    def check_is_usable_dataset(self, ds, items):
+        """ Check if a dataset is usable
+        """
+
+    def test_get_missing_dataset(self):
+        filename = os.path.join(self.temp_dir, 'test.cuds')
+        with closing(H5CUDS.open(filename)) as handle:
```
Comment (itziakos) [inline simphony/testing/abc_check_engine.py]: This template should be independent of the `H5CUDS` class.

**Comment by itziakos** [comment]
@roigcarlo the test template needs some work since it uses the `H5CUDS` instance. It could also make the template simpler if we provide utility methods to compare the different CUDS containers between them (remove more of the check_... methods).

**Comment by itziakos** [comment]
I would suggest to make the changes in this branch and reopen the PR.

**Comment by itziakos** [comment]
> Problem is that I didn't press the merge button, it was a typo in the terminal command...

That means that you have to wear a funny hat until the changes and fixes are accepted. :)

**Pull request merged**